src/net_processing.cpp

```diff
@@ -2392,6 +2392,20 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
                 }
                 pindexWalk = pindexWalk->pprev;
             }
+            // Special case for second cmpctblock request of tip
+            // Must match all conditions to add to vToFetch above (except mmapBlocksInFlight.count)
+            // and conditions to do CMPCT_BLOCK announcement below.
+            if (vToFetch.size() == 0 && pindexLast->pprev == chainActive.Tip() &&
+                mmapBlocksInFlight.size() == mmapBlocksInFlight.count(pindexLast->GetBlockHash()) &&
+                mmapBlocksInFlight.count(pindexLast->GetBlockHash()) < MAX_CMPCTBLOCKS_INFLIGHT_PER_BLOCK &&
+                !(pindexLast->nStatus & BLOCK_HAVE_DATA) &&
+                (!IsWitnessEnabled(pindexLast->pprev, chainparams.GetConsensus()) || State(pfrom->GetId())->fHaveWitness) &&
+                nodestate->fSupportsDesiredCmpctVersion) {
+                vToFetch.push_back(pindexLast);
```

Comment:
Further, simply adding the entry to vToFetch here may result in a null pointer dereference, I believe. If a peer announces two headers messages back-to-back, the first time we will MarkBlockAsInFlight to them, and the second time we'll hit this condition and add the entry to vToFetch again (which should also be fixed). Further down, we'll MarkBlockAsInFlight to them again, but MarkBlockAsInFlight requires that, if the block is already in-flight to the same peer, pit be something non-NULL as it will be dereferenced, but it is NULL in the call below.